### PR TITLE
Add gemfile flag to exec command

### PR DESCRIPTION
**Pull request by ankitkataria**
### What was the end-user problem that led to this PR?

This PR resolves the issue #5924 

### What was your diagnosis of the problem?

- added a `gemfile` flag in `cli.rb`
- if `options[:gemfile]` is set I created a SharedHelper method `custom_gemfile` which used the predefined `find_file` function to find and set the environment variable with the argument provided

### CODE REVIEW

**Comment by blackgurl1981** [comment]
Hi

**Comment by colby-swandale** [comment]
This needs a test to make sure the feature is behaving as expected and to make sure it will continue to work when new changes are made.

**Comment by ankitkataria** [comment]
@colby-swandale I'll start working on the test. Will update the pr asap. :wink:

File: spec/commands/exec_spec.rb [hunk 1]
```diff
@@ -6,6 +6,16 @@
...
   end
 
+  it "works with --gemfile flag", :focus do
+    create_file "CustomGemfile", <<-G
+      gem "rack"
+    G
+
+    bundle "exec --gemfile CustomGemfile test"
+
+    expect(ENV["BUNDLE_GEMFILE"]).to eq("CustomGemfile")
```
Comment (segiddins) [inline spec/commands/exec_spec.rb]: I dont think this will test what you want it to, because the `bundle` helper actually runs bundler in a subprocess

**Comment by ankitkataria** [comment]
@segiddins Thankyou so much for your comment :smile: I understood what I was doing wrong. I've updated the test now.

File: lib/bundler/cli/exec.rb [hunk 2]
```diff
@@ -23,6 +23,10 @@ def initialize(options, args)
     def run
       validate_cmd!
       SharedHelpers.set_bundle_environment
+
+      # setting the gemfile if mentioned
```
Comment (colby-swandale) [inline lib/bundler/cli/exec.rb]: This comment doesn't seem all that useful, the code already explains this

**Comment by colby-swandale** [comment]
Great! Thank you! 🎉 

@bundlerbot r+

**Comment by ankitkataria** [comment]
@colby-swandale I don't understand why the jobs corresponding to the failing tests were canceled. What needs to be done? :thinking:

File: spec/commands/exec_spec.rb [hunk 3]
```diff
@@ -6,6 +6,22 @@
     system_gems(system_gems_to_install, :path => :bundle_path)
   end
 
+  it "works with --gemfile flag" do
+    create_file "CustomGemfile", <<-G
```
Comment (colby-swandale) [inline spec/commands/exec_spec.rb]: Thinking more about this test, do you need 2 different Gemfiles?

Comment (ankitkataria) [reply spec/commands/exec_spec.rb]: I did that to ensure that the environment variable was set correctly both the times and `rackup` produced the version corresponding to the mentioned gemfiles.

File: lib/bundler/shared_helpers.rb [hunk 4]
```diff
@@ -230,6 +230,11 @@ def write_to_gemfile(gemfile_path, contents)
       filesystem_access(gemfile_path) {|g| File.open(g, "w") {|file| file.puts contents } }
     end
 
+    def custom_gemfile(gemfile)
```
[the remark this answers is not in the document]
Comment (ankitkataria) [reply lib/bundler/shared_helpers.rb]: It was mentioned in the issue that the flag could be extended to other commands as well for consistency. So I made a shared helper which could be used later. 
I used "custom" in the sense that it was mentioned by the user.  if it is not mentioned then the default gemfiles will be used.

[the remark this answers is not in the document]
Comment (ankitkataria) [reply lib/bundler/shared_helpers.rb]: @colby-swandale does this explanation make sense? is there a requirement to change the name of the function? :thinking:

File: spec/commands/exec_spec.rb (hunk 3 above)
Comment (colby-swandale) [reply spec/commands/exec_spec.rb]: You're already testing `bundle exec` with the `--gemfile` option. So testing it again is redundant, can you please remove it.

File: lib/bundler/shared_helpers.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (colby-swandale) [reply lib/bundler/shared_helpers.rb]: You should be able to use the custom Gemfile logic in the [CLI](https://github.com/bundler/bundler/blob/master/lib/bundler/cli.rb#L36) unless i'm missing something?

**Comment by colby-swandale** [comment]
Thanks! @bundlerbot r+

**Comment by colby-swandale** [comment]
@budlerbot retry

**Comment by colby-swandale** [comment]
@bundlerbot retry

**Pull request merged**